internal/wasm/memory.go

```diff
@@ -284,3 +292,49 @@ func (m *MemoryInstance) writeUint64Le(offset uint32, v uint64) bool {
 	binary.LittleEndian.PutUint64(m.Buffer[offset:], v)
 	return true
 }
+
+// Wait suspends the caller until the offset is notified by a different agent.
+func (m *MemoryInstance) Wait(offset uint32, timeout int64) (bool, bool) {
+	m.Mux.Lock()
+
+	if m.waiters == nil {
+		m.waiters = make(map[uint32][]chan struct{})
+	}
+
+	if uint(len(m.waiters[offset])) == math.MaxUint32 {
+		m.Mux.Unlock()
+		return true, false
+	}
+
+	ready := make(chan struct{})
+	m.waiters[offset] = append(m.waiters[offset], ready)
+	m.Mux.Unlock()
+
+	if timeout < 0 {
+		<-ready
+		return false, false
+	} else {
+		select {
+		case <-ready:
+			return false, false
+		case <-time.After(time.Duration(timeout)):
+			return false, true
+		}
+	}
+}
+
+// Notify wakes up at most count waiters at the given offset.
+func (m *MemoryInstance) Notify(offset uint32, count uint32) uint32 {
+	m.Mux.Lock()
+	defer m.Mux.Unlock()
+
+	res := uint32(0)
+	for num := len(m.waiters[offset]); num > 0 && res < count; num = len(m.waiters[offset]) {
+		w := m.waiters[offset][num-1]
+		m.waiters[offset] = m.waiters[offset][:num-1]
```

Comment:
This seems to notify in LIFO order.
I'm left wondering about liveness issues?

Reply:
Thanks for the reminder! Had intended to revisit but forgot, indeed LIFO seems like a bad idea for it (Wasm spec doesn't specify any order though). I switched to `list.List`, which is the same as the Go sempahore implementation